File: src/Graph.cpp
#include "Graph.hpp"

void Node::addEdge(Node* target, double delay, std::string type) {
    edges.push_back({target, delay, type});
}

Graph::~Graph() {
    for (auto const& [name, node] : nodes) {
        delete node;
    }
}

Node* Graph::getOrCreateNode(const std::string& name, const std::string& type) {
    if (nodes.find(name) == nodes.end()) {
        nodes[name] = new Node(name, type);
    }
    return nodes[name];
}

Node* Graph::getNode(const std::string& name) {
    if (nodes.find(name) != nodes.end()) {
        return nodes[name];
    }
    return nullptr;
}

std::vector<Node*> Graph::getAllNodes() {
    std::vector<Node*> all_nodes;
    for (auto const& [name, node] : nodes) {
        all_nodes.push_back(node);
    }
    return all_nodes;
}

void Graph::printSummary() {
    std::cout << "Graph Summary: Total Nodes = " << nodes.size() << std::endl;
}

File: sta_engine/graph.py
class Node:
    """Represents a pin or port in the timing graph."""
    def __init__(self, name, node_type):
        self.name = name # e.g., "u1.A" or "clk"
        self.type = node_type # "port", "pin"
        self.edges = [] # List of (target_node, weight, edge_type)
        self.at = -1.0 # Arrival Time
        self.rt = 999.0 # Required Time

    def add_edge(self, target, weight, edge_type):
        self.edges.append((target, weight, edge_type))

    def __repr__(self):
        return f"Node({self.name})"

class Graph:
    """Manages the graph nodes and connectivity."""
    def __init__(self):
        self.nodes = {} # name -> Node

    def get_or_create_node(self, name, node_type="pin"):
        if name not in self.nodes:
            self.nodes[name] = Node(name, node_type)
        return self.nodes[name]
    
    def get_node(self, name):
        return self.nodes.get(name)

    def get_all_nodes(self):
        return self.nodes.values()
    
    def summary(self):
        return f"Total Nodes: {len(self.nodes)}"
